### PR TITLE
Move more services to Library And enable real deletes

### PR DESCRIPTION
**Pull request by ryuyu**
Moves more of the implementations of required services into NuGetGallery.Services library.
Updates AccountDeleter to consume these changes.

Merge changes and test fixes coming soon.

Please look at changes by commit to make this more managable. Thanks.

### CODE REVIEW

File: src/AccountDeleter/Telemetry/GalleryTelemetryService.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public void TrackABTestEvaluated(string name, bool isActive, bool isAuthenticate
 
         public void TrackAccountDeletionCompleted(User deletedUser, User deletedBy, bool success)
         {
-            throw new NotImplementedException();
+            //no-op for now
```
Comment (shishirx34) [inline src/AccountDeleter/Telemetry/GalleryTelemetryService.cs]: is there a task to add telemetry here? may be add the TODO link to task

Comment (ryuyu) [reply src/AccountDeleter/Telemetry/GalleryTelemetryService.cs]: I switched us over in the newest commit to using the real gallery telemetry service, and have deleted by dev stub.

File: src/AccountDeleter/EmptyIndexingService.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Services.Entities;
+
+namespace NuGetGallery.AccountDeleter
+{
+    public class EmptyIndexingService : IIndexingService
+    {
+        public EmptyIndexingService()
```
Comment (cristinamanum) [inline src/AccountDeleter/EmptyIndexingService.cs]: Is this needed?

File: src/AccountDeleter/Evaluators/BaseUserEvaluator.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using NuGet.Services.Entities;
+
+namespace NuGetGallery.AccountDeleter
+{
+    public abstract class BaseUserEvaluator : IUserEvaluator
+    {
+        protected readonly Guid _id;
```
Comment (cristinamanum) [inline src/AccountDeleter/Evaluators/BaseUserEvaluator.cs]: It does not need to be protected. Private would be fine right?

File: src/AccountDeleter/GalleryAccountManager.cs [hunk 4]
```diff
@@ -56,6 +56,7 @@ public async Task<bool> DeleteAccount(User user)
             else
             {
                 _logger.LogError("Criteria passed but delete failed.");
+                _logger.LogError(result.Description);
```
Comment (cristinamanum) [inline src/AccountDeleter/GalleryAccountManager.cs]: Do we record metrics for users deleted and users failed to be deleted?

File: src/AccountDeleter/Job.cs [hunk 5]
```diff
@@ -57,9 +53,10 @@ protected override void ConfigureJobServices(IServiceCollection services, IConfi
             services.AddTransient<ISubscriptionProcessorTelemetryService, AccountDeleteTelemetryService>();
 
             services.AddScoped<AggregateEvaluator>();
-            services.AddTransient<AlwaysRejectEvaluator>();
-            services.AddTransient<AlwaysAllowEvaluator>();
-            services.AddTransient<UserPackageEvaluator>();
+            services.AddScoped<AlwaysRejectEvaluator>();
```
Comment (cristinamanum) [inline src/AccountDeleter/Job.cs]: Why Scoped ? Maybe add comments?

File: src/AccountDeleter/Job.cs [hunk 6]
```diff
@@ -120,17 +114,56 @@ protected void ConfigureGalleryServices(IServiceCollection services, IConfigurat
             }
             else
             {
-                services.AddScoped<IDeleteAccountService, EmptyDeleteAccountService>();
-                //services.AddScoped<IDeleteAccountService, DeleteAccountService>();
+                services.AddScoped<IEntitiesContext>(sp =>
+                {
+                    var connectionFactory = sp.GetRequiredService<ISqlConnectionFactory<GalleryDbConfiguration>>();
+                    var connection = connectionFactory.CreateAsync().GetAwaiter().GetResult();
```
Comment (cristinamanum) [inline src/AccountDeleter/Job.cs]: Why not await connectionFactory.CreateAsync() ?

File: src/NuGetGallery.Services/ServicesConstants.cs [hunk 7]
```diff
@@ -33,6 +35,17 @@ public static class ServicesConstants
         public const string EmailValidationRegex = "^" + EmailValidationRegexFirstPart + "@" + EmailValidationRegexSecondPart + "$";
         public static TimeSpan EmailValidationRegexTimeout = TimeSpan.FromSeconds(5);
 
+        /// <summary>
+        /// Parameters for calculating account lockout period after 
+        /// wrong password entry.
+        /// </summary>
+        public const double AccountLockoutMultiplierInMinutes = 10;
+        public const double AllowedLoginAttempts = 10;
```
Comment (cristinamanum) [inline src/NuGetGallery.Services/ServicesConstants.cs]: Is this new ? I do not see it removed from other file ?

File: src/AccountDeleter/Evaluators/BaseUserEvaluator.cs (hunk 3 above)
Comment (ryuyu) [reply src/AccountDeleter/Evaluators/BaseUserEvaluator.cs]: Yes i think you are right. Private would be fine since we can use EvaluatorId in derived classes.

File: src/AccountDeleter/Job.cs (hunk 6 above)
Comment (ryuyu) [reply src/AccountDeleter/Job.cs]: I copied this from the way we hydrate these connections in our other jobs. I don't think i can use await here as it would force the rest of everything here to become asynchronous, which i don't think we currently support.

File: src/AccountDeleter/EmptyIndexingService.cs (hunk 2 above)
Comment (ryuyu) [reply src/AccountDeleter/EmptyIndexingService.cs]: One of the guys that we are hydrating with DI requires an IIndexingService in order, but we don't use this in our current system since we index over V3 which directly reads DB to its updates. This is so the service that requests this can make the calls it wants, but we hijack them and no-op.

File: src/AccountDeleter/GalleryAccountManager.cs (hunk 4 above)
Comment (ryuyu) [reply src/AccountDeleter/GalleryAccountManager.cs]: Yes we track an event using TrackDeleteResult. It just happens where this gets called above in the MessageHandler.

File: src/NuGetGallery.Services/ServicesConstants.cs (hunk 7 above)
Comment (ryuyu) [reply src/NuGetGallery.Services/ServicesConstants.cs]: These should have been removed from GalleryConstants. I have updated this. Good catch.

File: src/AccountDeleter/EmptyIndexingService.cs (hunk 2 above)
Comment (cristinamanum) [reply src/AccountDeleter/EmptyIndexingService.cs]: Maybe comments.

File: src/AccountDeleter/Configuration/GalleryConfiguration.cs [hunk 8]
```diff
@@ -10,6 +10,16 @@ namespace NuGetGallery.AccountDeleter
 {
     public class GalleryConfiguration : IAppConfiguration
     {
+        public string SiteRoot
+        {
+            get
+            {
+                return "";
+            }
```
Comment (shishirx34) [inline src/AccountDeleter/Configuration/GalleryConfiguration.cs]: why return empty string?

Comment (ryuyu) [reply src/AccountDeleter/Configuration/GalleryConfiguration.cs]: This guy is called by someone during a constructor, but as of today, AccountDeleter doesn't care about this value. However, the guy that calls it expects to get a non-null result.

File: src/AccountDeleter/Evaluators/UserOrganizationEvaluator.cs [hunk 9]
```diff
@@ -10,30 +10,19 @@ namespace NuGetGallery.AccountDeleter
...
-            get
-            {
-                return _id.ToString();
-            }
-        }
-
-        public bool CanUserBeDeleted(User user)
+        public override bool CanUserBeDeleted(User user)
         {
             throw new NotImplementedException();
```
Comment (shishirx34) [inline src/AccountDeleter/Evaluators/UserOrganizationEvaluator.cs]: what are the plans to use this evaluator? Is it needed?

File: src/AccountDeleter/GalleryAccountManager.cs (hunk 4 above)
Comment (shishirx34) [reply src/AccountDeleter/GalleryAccountManager.cs]: is the metric tracking inside the `DeleteAccountAsync` method?

File: src/AccountDeleter/Job.cs (hunk 5 above)
Comment (shishirx34) [reply src/AccountDeleter/Job.cs]: Also, don't see the AccountOrganization evaluator, is that needed to be included anywhere? may be eventually?

File: src/AccountDeleter/Configuration/AccountDeleteConfiguration.cs [hunk 10]
```diff
@@ -28,6 +28,11 @@ public class AccountDeleteConfiguration
         /// </summary>
         public Dictionary<string, string> TemplateReplacements { get; set; }
 
+        /// <summary>
+        /// Storage container connection string where gallery content can be found
+        /// </summary>
+        public string GalleryStorageConnectionString { get; set; }
```
Comment (skofman1) [inline src/AccountDeleter/Configuration/AccountDeleteConfiguration.cs]: why is this needed?

Comment (ryuyu) [reply src/AccountDeleter/Configuration/AccountDeleteConfiguration.cs]: Feature flags clean up requires access to the gallery blob where it is held.

File: src/AccountDeleter/EmptyIndexingService.cs (hunk 2 above)
Comment (shishirx34) [reply src/AccountDeleter/EmptyIndexingService.cs]: I see `PackageUpdateService` uses it. I know in gallery we used the `LuceneIndexingService` as the default, doesn't sound like it should be moved here. 👍 for this empty service.

File: src/AccountDeleter/Job.cs [hunk 11]
```diff
@@ -120,17 +114,56 @@ protected void ConfigureGalleryServices(IServiceCollection services, IConfigurat
...
+                    var optionsSnapshot = options.Value;
+
+                    return new CloudBlobClientWrapper(optionsSnapshot.GalleryStorageConnectionString, readAccessGeoRedundant: true);
+                });
 
-                services.AddScoped<ITelemetryService, GalleryTelemetryService>();
+                services.AddScoped<ITelemetryService, TelemetryService>();
                 services.AddScoped<ISecurityPolicyService, SecurityPolicyService>();
-                services.AddScoped<IAuditingService>(sp => { return AuditingService.None; }); //Replace with real when we start doing deletes. For now, we are a readonly operation.
+                services.AddScoped<IAuditingService>(sp => { return AuditingService.None; });
```
Comment (shishirx34) [inline src/AccountDeleter/Job.cs]: I would leave the comment, doesn't look like we have auditing integrated yet?

Comment (shishirx34) [reply src/AccountDeleter/Job.cs]: unless auditing is being tracked with an issue for deletions.

File: src/AccountDeleter/Configuration/GalleryConfiguration.cs (hunk 8 above)
Comment (agr) [reply src/AccountDeleter/Configuration/GalleryConfiguration.cs]: Maybe use expression body, like in the setter below? `get => "";`

File: src/AccountDeleter/Evaluators/UserOrganizationEvaluator.cs (hunk 9 above)
Comment (ryuyu) [reply src/AccountDeleter/Evaluators/UserOrganizationEvaluator.cs]: Might not be needed. We will be discussing exactly what evaluations we want to run in the next few weeks. At that point, we will either complete this implementation, or remove to evaluator altogether.

File: src/AccountDeleter/Job.cs (hunk 11 above)
Comment (ryuyu) [reply src/AccountDeleter/Job.cs]: According to @skofman1, we don't have this auditing in this path. If you know otherwise, let's file an issue and link here and we can move that too.
